sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -493,27 +494,29 @@ mod checked {
         // For newly published packages, runtime ID matches storage ID.
         let storage_id = runtime_id;
         let dependencies = fetch_packages(context, &dep_ids)?;
-        let package_obj = context.new_package(&modules, &dependencies)?;
-
-        let Some(package) = package_obj.data.try_as_package() else {
-        invariant_violation!("Newly created package object is not a package");
-    };
-
-        context.set_linkage(package)?;
+        let package = context.new_package(&modules, &dependencies)?;
+
+        // Here we optimistacally push the package that is being published/upgraded
+        // and if there is an error of any kind (verification or module init) we
+        // remove it.
+        // The call to `pop_last_package` later is fine because we cannot re-enter and
+        // the last package we pushed is the one we are verifying and running the init from
```

Comment:
this is the part I mentioned in the comment about in `pop_last_package`.
Let me know if you are fine with it, or you want to throw a tantrum...